README.md
# Deep Learning based network intrusion detection in PyTorch

Net intrusion detection experiment for Final Project of DeepLearning class at Inha University.

## Dataset homepage: https://www.unb.ca/cic/datasets/ids-2017.html

## Contributions

1. Correct Evaluation Metric
2. Adressing data imblance
3. Benchmark results for different ML models
4. Running code for training/evaluating

## Accompanying slides

https://docs.google.com/presentation/d/1Rjj1vF0hv8vSJWeDxk23nE4A4w3fv8tBdvsyIBpWTdU/edit?usp=sharing

## Model Performance using K-Fold Cross-Validation

| Classifier                        | 1-Fold Balanced Accuracy | 5-Fold Balanced Accuracy |
| --------------------------------- | ------------------------ | ------------------------ |
| Content Linear Softmax            | 80.99                    | 76.27                    |
| Neural Network with 3 dense layer | 84.87                    | 85.73                    |
| Neural Network with 5 dense layer | 85.57                    | 85.63                    |
| 1D-CNN with 2conv 1fc layer       | 87.11                    | 87.13                    |
| CNN with 5conv layer              | 83.29                    | 87.16                    |
| Random Forest                     | 80.69                    | 80.09                    |

### Softmax

Please run the Softmax.ipynb

### NN

Please run the NN.ipynb
There are two NN architectures:

1. 'nn3' - 3 layers
2. 'nn5' - 5 layers

### 1D-CNN

Please run the CNN.ipynb
There are two 1D-CNN architectures:

1. 'cnn2' - 2 conv layers
2. 'cnn5' - 5 conv layers


releases.md
# Release Notes

All notable changes to the machine learning experiment repository with models nn3, nn5, conv, and randomforest on the IDS2017 dataset will be documented in this file.

## [Version 1.0.1] - 2023-05-18

### Added

- Initial release of the repository.
- Modified dataset download process to use a Google Drive link for faster downloading.
- Added logging functionality to save running progress instead of using direct print statements.
- Updated existing codebase on models.py with the `Method` class, including the following methods:
  - `SOFTMAX`: Softmax method for classification.
  - `CNN2`: Convolutional Neural Network with 2 layers.
  - `CNN5`: Convolutional Neural Network with 5 layers.
  - `NN3`: Neural Network with 3 layers.
  - `NN5`: Neural Network with 5 layers.

### Fixed

- Removed unnecessary cells and code snippets from the Jupyter Notebook files.

### Security

- No security updates in this release.


test_datamodule.py
import argparse

import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader

from dataset import MyDataModule


def main(args):
    # Initialize data module
    data_module = MyDataModule(
        dataroot=args.dataroot,
        batch_size=args.batch_size,
        seed=args.seed,
    )
    data_module.prepare_data()
    data_module.setup()

    # Print dataset shapes
    print(f"Train inputs shape: {data_module.train_dataset.X.shape}")
    print(f"Train targets shape: {data_module.train_dataset.y.shape}")
    print(f"Validation inputs shape: {data_module.val_dataset.X.shape}")
    print(f"Validation targets shape: {data_module.val_dataset.y.shape}")
    print(f"Test inputs shape: {data_module.test_dataset.X.shape}")
    print(f"Test targets shape: {data_module.test_dataset.y.shape}")

    # Initialize data loader
    dataloader = DataLoader(
        data_module.train_dataset,
        batch_size=args.batch_size,
        shuffle=True,
        num_workers=4,
        pin_memory=True,
    )

    # Print batch shapes
    inputs, targets = next(iter(dataloader))
    print(f"Batch inputs shape: {inputs.shape}")
    print(f"Batch targets shape: {targets.shape}")

# python test_datamodule.py MachineLearningCVE/
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Test MyDataModule")
    parser.add_argument("dataroot", type=str, help="path to dataset root directory")
    parser.add_argument("--batch-size", type=int, default=32, help="batch size for training")
    parser.add_argument("--seed", type=int, default=42, help="random seed")
    args = parser.parse_args()
    main(args)


cnn.py
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision.transforms as transforms
import torch.utils.data as utils
from torch.utils.tensorboard import SummaryWriter
from sklearn import metrics
from os.path import join


class CNN2(nn.Module):

    def __init__(self,input_dim,num_classes,use_batchnorm=False):
        super(CNN2, self).__init__()
        # kernel
        self.input_dim = input_dim 
        self.num_classes = num_classes
        self.use_batchnorm = use_batchnorm
        
        conv_layers = []
        
        conv_layers.append(nn.Conv1d(in_channels=1,out_channels=64,kernel_size=3,padding=1)) # ;input_dim,64
        conv_layers.append(nn.BatchNorm1d(64))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=64,out_channels=128,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(128))
        conv_layers.append(nn.ReLU(True))
        
        self.conv = nn.Sequential(*conv_layers)

        fc_layers = []
        fc_layers.append(nn.Linear(input_dim*128,num_classes))
        self.classifier = nn.Sequential(*fc_layers)
        
    def forward(self, x):
        x = self.conv(x)
        x = torch.flatten(x,1)
        x = self.classifier(x)
        return x




class CNNClassifier():
    def __init__(self, input_dim,num_classes,num_epochs=5,batch_size=100,lr=5e-3,reg=2.5e-2,runs_dir=None,use_batchnorm=False):
        self.model = CNN2(input_dim,num_classes,use_batchnorm)
        self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        self.model = self.model.to(self.device)

        self.num_epochs = num_epochs
        self.num_classes = num_classes
        self.batch_size = batch_size
        self.lr = lr
        self.reg = reg
        self.runs_dir = runs_dir

        self.criterion = nn.CrossEntropyLoss()
        self.optimizer = torch.optim.Adam(self.model.parameters(),lr=lr,betas=(0.9,0.99),eps=1e-08, weight_decay=reg, amsgrad=False)

  
    def fit(self,x,y,x_val,y_val,verbose=True):
        writer = SummaryWriter(self.runs_dir) 

        tensor_x = torch.stack([torch.Tensor(i) for i in x]).to(self.device)
        tensor_y = torch.LongTensor(y).to(self.device) # checked working correctly

        dataset = utils.TensorDataset(tensor_x,tensor_y)
        train_loader = utils.DataLoader(dataset,batch_size=self.batch_size) 
        model = self.model
        best_acc = -1
        for epoch in range(self.num_epochs):
            for i,(xi,yi) in enumerate(train_loader):
                # xi (bs,D)
                batch_size, D = xi.shape
                xi = xi.view(batch_size,1,D)
                
                outputs = model(xi)
                loss = self.criterion(outputs,yi)
                seen_so_far = self.batch_size*(epoch*len(train_loader)+i+1) # fixes issues with different batch size
                writer.add_scalar('Loss/train',loss.item(),seen_so_far)
                #batckward, optimize
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
                if i%50==0 :
                    raw_pred,pred = self.predict(x_val)
                    balanced_acc = metrics.balanced_accuracy_score(y_val,pred)*100
                    if balanced_acc>best_acc:
                        best_acc = balanced_acc
                        
                        checkpoint = {
                        'state_dict': model.state_dict(),
                        'optimizer' : self.optimizer.state_dict(),
                        'epoch':epoch,
                        'batch': i,
                        'batch_size': self.batch_size
                        }
                        self.save(checkpoint) 
                    writer.add_scalar('Accuracy/Balanced Val',balanced_acc,seen_so_far)

                    acc = metrics.accuracy_score(y_val,pred)*100
                    writer.add_scalar('Accuracy/Val',acc,seen_so_far)

                    if verbose:
                        print ('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}' 
                                               .format(epoch+1, self.num_epochs, i+1, len(y)//self.batch_size, loss.item()))
        writer.close()

    def predict(self,x,eval_mode=False):
        tensor_x = torch.stack([torch.Tensor(i) for i in x]).to(self.device)
        bs = self.batch_size
        num_batch = x.shape[0]//bs +1*(x.shape[0]%bs!=0)

        pred = torch.zeros(0,dtype=torch.int64).to(self.device)
        raw_pred = torch.zeros((0,self.num_classes)).to(self.device)
        if eval_mode:
            model = self.load_checkpoint()
        else:
            model = self.model
        model.eval()

        with torch.no_grad():
            for i in range(num_batch):
                xi = tensor_x[i*bs:(i+1)*bs]
                batch_size, D = xi.shape
                xi = xi.view(batch_size,1,D)
                outputs = model(xi) # N,C
                _, predi = torch.max(outputs.data,1)
                pred = torch.cat((pred,predi))
                raw_pred = torch.cat((raw_pred,outputs.data))
        return raw_pred.cpu().numpy(),pred.cpu().numpy()

    def save(self,checkpoint):
        path = join(self.runs_dir,'checkpoint.pth')
        torch.save(checkpoint,path)


    def load_checkpoint(self,inference_mode=True):
        filepath = join(self.runs_dir,'checkpoint.pth')
        checkpoint = torch.load(filepath)
        model = self.model
        model.load_state_dict(checkpoint['state_dict'])
        
        print("Loaded {} model trained with batch_size = {}, seen {} epochs and {} mini batches".
            format(self.runs_dir,checkpoint['batch_size'],checkpoint['epoch'],checkpoint['batch']))
    
        if inference_mode:
            for parameter in model.parameters():
                parameter.requires_grad = False
            model.eval()
        return model

dataset.py
import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader, Dataset
from sklearn.model_selection import train_test_split
from imblearn.combine import SMOTEENN
from preprocessing import load_data, balance_data
from sklearn.model_selection import StratifiedKFold, train_test_split
import numpy as np

class MyDataset(Dataset):
    """
    A custom Dataset class for containing samples and their labels.
    """
    def __init__(self, X, y):
        """
        Initialize the dataset with input features X and labels y.

        Args:
            X (array-like): Input feature data.
            y (array-like): Corresponding labels for the input data.
        """
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.long)

    def __len__(self):
        return len(self.X)
    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]
    # def __getitem__(self, idx):
    #     return self.X[idx], self.y[idx]

class MyDataModule(pl.LightningDataModule):
    """
    Custom data module for handling data processing, split, and creating data loaders.
    """
    def __init__(self, dataroot, batch_size, seed, n_splits=5):
        """
        Initialize the data module with the data root directory, batch size, seed for reproducibility, 
        and number of splits for cross-validation.

        Args:
            dataroot (str): The root directory containing the data.
            batch_size (int): Batch size for DataLoader.
            seed (int): Seed for random number generation to ensure reproducibility.
            n_splits (int, optional): Number of splits for cross-validation. Defaults to 5.
        """
        super().__init__()
        self.dataroot = dataroot
        self.batch_size = batch_size
        self.seed = seed
        self.n_splits = n_splits

    def prepare_data(self):
        # Load and preprocess data here, if necessary
        pass

    def setup(self, stage=None):
        X, y = load_data(self.dataroot)
        skf = StratifiedKFold(n_splits=self.n_splits, shuffle=True, random_state=self.seed)

        # Assuming you want the first fold for simplicity, you can change this logic to pick a specific fold
        train_val_indices, test_indices = next(skf.split(X, y))
        X_train_val, X_test = X[train_val_indices], X[test_indices]
        y_train_val, y_test = y[train_val_indices], y[test_indices]

        skf_inner = StratifiedKFold(n_splits=self.n_splits, shuffle=True, random_state=self.seed)

        # Assuming you want the first inner fold for simplicity, you can change this logic to pick a specific fold
        train_indices, val_indices = next(skf_inner.split(X_train_val, y_train_val))
        X_train, X_val = X_train_val[train_indices], X_train_val[val_indices]
        y_train, y_val = y_train_val[train_indices], y_train_val[val_indices]

        self.train_dataset = MyDataset(X_train, y_train)
        self.val_dataset = MyDataset(X_val, y_val)
        self.test_dataset = MyDataset(X_test, y_test)
        
        input_sample, _ = self.train_dataset[0]
        self.input_dim = len(input_sample)
        self.output_dim = len(torch.unique(torch.tensor(y_train)))
    @property
    def input_dim(self):
        return self._input_dim

    @input_dim.setter
    def input_dim(self, value):
        self._input_dim = value

    @property
    def output_dim(self):
        return self._output_dim

    @output_dim.setter
    def output_dim(self, value):
        self._output_dim = value

    def train_dataloader(self):
        return DataLoader(
            self.train_dataset,
            batch_size=self.batch_size,
            shuffle=True,
            num_workers=4,
            pin_memory=True,
        )

    def val_dataloader(self):
        return DataLoader(
            self.val_dataset,
            batch_size=self.batch_size,
            num_workers=4,
            pin_memory=True,
        )

    def test_dataloader(self):
        return DataLoader(
            self.test_dataset,
            batch_size=self.batch_size,
            num_workers=4,
            pin_memory=True,
        )



shallows.py
import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms
import torch.utils.data as utils
from torch.utils.tensorboard import SummaryWriter

class LinearClassifier:
    def __init__(self,lossfunction,input_size,num_classes,num_epochs=5,batch_size=100,lr=5e-3,reg=2.5e-2,runs_dir=None):
        self.input_size = input_size
        self.num_classes = num_classes
        self.num_epochs = num_epochs
        self.batch_size = batch_size
        self.learning_rate = lr
        self.reg= reg
        self.lossfunction = lossfunction
        self.runs_dir = runs_dir
        self.device='cpu'
        self.build()

    def build(self):

        self.model = nn.Linear(self.input_size, self.num_classes).to(self.device)
        print("Model is created")
        if self.lossfunction=='softmax':
            self.criterion = nn.CrossEntropyLoss()
        elif self.lossfunction=='svm':
            pass
        else:
            print('Please specify loss function')
            exit()
            
        self.optimizer = torch.optim.Adam(self.model.parameters(),lr=self.learning_rate,betas=(0.9,0.99),eps=1e-08, weight_decay=self.reg, amsgrad=False)

    def fit(self,x,y,verbose=True):
        writer = SummaryWriter(self.runs_dir) 

        tensor_x = torch.stack([torch.Tensor(i) for i in x]).to(self.device)
        tensor_y = torch.LongTensor(y).to(self.device) # checked working correctly
        dataset = utils.TensorDataset(tensor_x,tensor_y)
        train_loader = utils.DataLoader(dataset,batch_size=self.batch_size) 
        for epoch in range(self.num_epochs):
            for i,(xi,yi) in enumerate(train_loader):
                outputs = self.model(xi)
                loss = self.criterion(outputs,yi)
                seen_so_far = self.batch_size*(epoch*len(train_loader)+i+1) # fixes issues with different batch size
                writer.add_scalar('Loss/train',loss.item(),seen_so_far)
                #batckward, optimize
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
                if i%100==0 and verbose:
                    print ('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}' 
                                               .format(epoch+1, self.num_epochs, i+1, len(y)//self.batch_size, loss.item()))
        writer.close()

    def predict(self,x):
        tensor_x = torch.stack([torch.Tensor(i) for i in x]).to(self.device)
        bs = self.batch_size
        num_batch = x.shape[0]//bs +1*(x.shape[0]%bs!=0)

        pred = torch.zeros(0,dtype=torch.int64).to(self.device)
        with torch.no_grad():
            for i in range(num_batch):
                xi = tensor_x[i*bs:(i+1)*bs]
                outputs = self.model(xi)
                _, predi = torch.max(outputs.data,1)
                pred = torch.cat((pred,predi))
        
        return pred.cpu().numpy()




models.py
import torch
import torch.nn as nn
import torch.utils.data as utils
from torch.utils.tensorboard import SummaryWriter
import os
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn import metrics
from enum import Enum
import logging
from os.path import join


class Softmax(nn.Module):
    """ 
    Softmax Regression Model
    """
    def __init__(self,input_dim,num_classes,device):
        super(Softmax,self).__init__()
        self.classifier = nn.Linear(input_dim, num_classes).to(device)
        #self.classifier = self.classifier'

    def forward(self,x):
        output= self.classifier(x)         
        return output


class CNN2(nn.Module):
    """ 
    2-layer Convolutional Neural Network Model
    """
    def __init__(self,input_dim,num_classes,device):
        super(CNN2, self).__init__()
        # kernel
        self.input_dim = input_dim
        self.num_classes = num_classes

        conv_layers = []
        conv_layers.append(nn.Conv1d(in_channels=1,out_channels=64,kernel_size=3,padding=1)) # ;input_dim,64
        conv_layers.append(nn.BatchNorm1d(64))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=64,out_channels=128,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(128))
        conv_layers.append(nn.ReLU(True))

        self.conv = nn.Sequential(*conv_layers).to(device)

        fc_layers = []
        fc_layers.append(nn.Linear(input_dim*128,num_classes))
        self.classifier = nn.Sequential(*fc_layers).to(device)

    def forward(self, x):
        batch_size, D = x.shape
        x = x.view(batch_size,1,D)

        x = self.conv(x)
        x = torch.flatten(x,1)
        x = self.classifier(x)
        return x

class CNN5(nn.Module):
    """ 
    5-layer Convolutional Neural Network Model
    """
    def __init__(self,input_dim,num_classes,device):
        super(CNN5, self).__init__()
        # kernel
        self.input_dim = input_dim
        self.num_classes = num_classes

        conv_layers = []
        conv_layers.append(nn.Conv1d(in_channels=1,out_channels=64,kernel_size=3,padding=1)) # ;input_dim,64
        conv_layers.append(nn.BatchNorm1d(64))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=64,out_channels=128,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(128))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=128,out_channels=256,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(256))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=256,out_channels=256,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(256))
        conv_layers.append(nn.ReLU(True))

        conv_layers.append(nn.Conv1d(in_channels=256,out_channels=128,kernel_size=3,padding=1)) #(input_dim,128)
        conv_layers.append(nn.BatchNorm1d(128))
        conv_layers.append(nn.ReLU(True))
        
        self.conv = nn.Sequential(*conv_layers).to(device)

        fc_layers = []
        fc_layers.append(nn.Linear(input_dim*128,num_classes))
        self.classifier = nn.Sequential(*fc_layers).to(device)

    def forward(self, x):
        batch_size, D = x.shape
        x = x.view(batch_size,1,D)

        x = self.conv(x)
        x = torch.flatten(x,1)
        x = self.classifier(x)
        return x


class Net3(nn.Module):
    """
    A neural network model consisting of multiple layers, used for classification tasks.

    Args:
    - input_dim (int): The dimensionality of the input data.
    - num_classes (int): The number of classes in the classification problem.
    - device (str): The device to be used for running the computations.

    Attributes:
    - input_dim (int): The dimensionality of the input data.
    - num_classes (int): The number of classes in the classification problem.
    - device (str): The device to be used for running the computations.

    Methods:
    - forward(x): Defines the forward pass of the model, taking in an input tensor x and returning the output tensor.

    """
    def __init__(self,input_dim,num_classes,device):
        super(Net3, self).__init__()
        # kernel
        print('building NN3')
        self.input_dim = input_dim
        self.num_classes = num_classes

        layers = []
        layers.append(nn.Dropout(p=0.1))
        layers.append(nn.Linear(self.input_dim, 128))
        layers.append(nn.BatchNorm1d(num_features=128)) 
        layers.append(nn.Dropout(p=0.3))
        layers.append(nn.Linear(128, 128))
        layers.append(nn.BatchNorm1d(num_features=128))
        layers.append(nn.Linear(128, self.num_classes))
        self.classifier = nn.Sequential(*layers).to(device)
        
    def forward(self, x):
        x = self.classifier(x)
        return x


class Net5(nn.Module):
    """
    A neural network model with 4 fully connected layers, using batch normalization and dropout for regularization.

    Args:
    - input_dim (int): the number of input features
    - num_classes (int): the number of output classes
    - device (torch.device): the device on which to run the model

    Attributes:
    - input_dim (int): the number of input features
    - num_classes (int): the number of output classes
    - model (nn.Sequential): the neural network architecture consisting of 4 fully connected layers
    """
    def __init__(self,input_dim,num_classes,device):
        super(Net5, self).__init__()
        # kernel
        self.input_dim = input_dim
        self.num_classes = num_classes
        
        layers = []
        layers.append(nn.Linear(input_dim,128))

        layers.append(nn.BatchNorm1d(128))
        layers.append(nn.ReLU(True))
        layers.append(nn.Linear(128,256))
        
        layers.append(nn.BatchNorm1d(256))
        layers.append(nn.Dropout(p=0.3))
        layers.append(nn.ReLU(True))
        layers.append(nn.Linear(256,256))
        
        layers.append(nn.BatchNorm1d(256))
        layers.append(nn.Dropout(p=0.4))
        layers.append(nn.ReLU(True))
        layers.append(nn.Linear(256,128))

        layers.append(nn.BatchNorm1d(128))
        layers.append(nn.Dropout(p=0.5))
        layers.append(nn.ReLU(True))        
        layers.append(nn.Linear(128,num_classes))

        self.model = nn.Sequential(*layers).to(device)
        
    def forward(self, x):
        return self.model(x)

class Method(Enum):
    SOFTMAX = "softmax"
    CNN2 = "cnn2"
    CNN5 = "cnn5"
    NN3 = "nn3"
    NN5 = "nn5"

class Classifier:
    """
    A classifier for machine learning models using PyTorch.

    Parameters
    ----------
    method : str
        The method used for classification. Currently supported options are
        'softmax', 'cnn2', 'cnn5', 'nn3', and 'nn5'.
    input_dim : int
        The number of input features.
    num_classes : int
        The number of classes.
    num_epochs : int
        The number of training epochs.
    batch_size : int, optional
        The batch size. Default is 100.
    lr : float, optional
        The learning rate. Default is 1e-3.
    reg : float, optional
        The regularization strength. Default is 1e-5.
    runs_dir : str, optional
        The directory for saving training runs. Default is None.

    Attributes
    ----------
    batch_size : int
        The batch size.
    num_epochs : int
        The number of training epochs.
    learning_rate : float
        The learning rate.
    reg : float
        The regularization strength.
    runs_dir : str
        The directory for saving training runs.
    device : torch.device
        The device used for training.
    model : torch.nn.Module
        The PyTorch model used for classification.
    criterion : torch.nn.Module
        The PyTorch criterion used for optimization.
    optimizer : torch.optim.Optimizer
        The PyTorch optimizer used for training.

    Methods
    -------
    fit(X, Y)
        Trains the classifier on the input data X and labels Y.
    predict(x, eval_mode=False)
        Predicts the labels for the input data x.
    """
    def __init__(self,method,input_dim,num_classes,num_epochs,batch_size=100,lr=1e-3,reg=1e-5,runs_dir=None,seed=10):
        self.batch_size = batch_size
        self.num_epochs = num_epochs
        self.learning_rate = lr
        self.reg= reg
        self.runs_dir = runs_dir
        self.seed = seed
        #self.device = 'cuda'
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.DEBUG)
        file_handler = logging.FileHandler('training.log')
        file_handler.setLevel(logging.DEBUG)
        file_formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        file_handler.setFormatter(file_formatter)
        self.logger.addHandler(file_handler)
        self.logger.info("Classifier initialized with method %s, input_dim %d, num_classes %d, num_epochs %d, batch_size %d, lr %f, reg %f, runs_dir %s" % (method,input_dim,num_classes,num_epochs,batch_size,lr,reg,runs_dir))
        
        #self.model = nn.Linear(self.input_size, self.num_classes).to(self.device)
        if method==Method.SOFTMAX.value:
            self.device = torch.device('cuda:1')
            self.model = Softmax(input_dim,num_classes=num_classes, device=self.device)
        elif method==Method.CNN2.value:
            self.device = torch.device('cuda:2')
            self.model = CNN2(input_dim,num_classes=num_classes,device=self.device)        
        elif method==Method.CNN5.value:
            self.device = torch.device('cuda:1')
            self.model = CNN5(input_dim,num_classes=num_classes,device=self.device)        
        elif method==Method.NN3.value:
            self.device = torch.device('cuda:0')
            self.model = Net3(input_dim,num_classes=num_classes,device=self.device)        
        elif method==Method.NN5.value:
            self.device = torch.device('cuda:0')
            self.model = Net5(input_dim,num_classes=num_classes,device=self.device)        
        else:
            raise ValueError("Method must be one of 'softmax', 'cnn2', 'cnn5', 'nn3', or 'nn5'.")
        
        self.criterion = nn.CrossEntropyLoss()
        self.optimizer = torch.optim.Adam(self.model.parameters(),lr=self.learning_rate,betas=(0.9,0.99),eps=1e-08, weight_decay=self.reg, amsgrad=False)

    def fit(self,X,Y):
        self.logger.info('Starting training process...')
        sss = StratifiedShuffleSplit(n_splits=1, test_size=0.1, random_state=self.seed)
        for dev_index, val_index in sss.split(X, Y): # runs only once
                X_dev = X[dev_index]
                Y_dev = Y[dev_index]
                X_val = X[val_index]
                Y_val = Y[val_index]  
        
        writer = SummaryWriter(self.runs_dir) 

        tensor_x = torch.stack([torch.Tensor(i) for i in X_dev]).to(self.device)
        tensor_y = torch.LongTensor(Y_dev).to(self.device) # checked working correctly

        dataset = utils.TensorDataset(tensor_x,tensor_y)        
        train_loader = utils.DataLoader(dataset,batch_size=self.batch_size) 
        N = tensor_x.shape[0]

        num_epochs = self.num_epochs

        model  = self.model
        best_acc = None
        best_epoch = None

        filepath = join(self.runs_dir,'checkpoint.pth')
        if os.path.isfile(filepath):
            checkpoint = self.load_checkpoint(filepath)
            best_epoch = checkpoint['epoch']
            best_batch = checkpoint['batch']
            self.model.load_state_dict(checkpoint['state_dict'])
            self.optimizer.load_state_dict(checkpoint['optimizer'])
            pred = self.predict(X_val)
            best_acc = metrics.balanced_accuracy_score(Y_val,pred)*100
            resume_epoch = best_epoch+1  
            resume_batch = best_batch+1
        else:
            resume_epoch = 0
            resume_batch = 0
            best_acc = -1
            best_epoch = 0

        no_improvement = 0
        for epoch in range(resume_epoch,num_epochs):
            for i,(xi,yi) in enumerate(train_loader):
                if epoch==resume_epoch and i<resume_batch:
                    continue
                    
                outputs = model(xi)
                loss = self.criterion(outputs,yi)

                loss.requires_grad
                seen_so_far = self.batch_size*(epoch*len(train_loader)+i+1) # fixes issues with different batch size
                writer.add_scalar('Loss/train',loss.item(),seen_so_far)
                #batckward, optimize
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()
                
                if (seen_so_far/self.batch_size)%50==0:
                    pred = self.predict(X_val)
                    balanced_acc = metrics.balanced_accuracy_score(Y_val,pred)*100
                    if balanced_acc > best_acc:
                        best_acc = balanced_acc
                        best_epoch = epoch
                        checkpoint = {
                        'state_dict': model.state_dict(),
                        'optimizer' : self.optimizer.state_dict(),
                        'epoch':epoch,
                        'batch': i,
                        'batch_size': self.batch_size
                        }
                        self.save(checkpoint)
                        no_improvement =0
                    else:
                        no_improvement+=1
                        if no_improvement>=10:
                            self.logger.warning("No improvement in accuracy for 10 iterations.")
                            return
                    self.logger.debug('Epoch [%d/%d], Step [%d/%d], Loss: %.4f', epoch+1, num_epochs, i+1, len(Y_dev)//self.batch_size, loss.item())


                    writer.add_scalar('Accuracy/Balanced Val',balanced_acc,seen_so_far)

                    acc = metrics.accuracy_score(Y_val,pred)*100
                    writer.add_scalar('Accuracy/Val',acc,seen_so_far)
        writer.close()

    def predict(self,x,eval_mode=False):
        tensor_x = torch.stack([torch.Tensor(i) for i in x]).to(self.device)
        bs = self.batch_size
        num_batch = x.shape[0]//bs +1*(x.shape[0]%bs!=0)

        pred = torch.zeros(0,dtype=torch.int64).to(self.device)
        
        if eval_mode:
            model = self.load_model()
        else:
            model = self.model
        model.eval()        
        
        with torch.no_grad():
            for i in range(num_batch):
                xi = tensor_x[i*bs:(i+1)*bs]
                outputs = model(xi)
                _, predicted_labels = torch.max(outputs.data,1)
                pred = torch.cat((pred,predicted_labels))

        return pred.cpu().numpy()


    def save(self,checkpoint):
        path = join(self.runs_dir,'checkpoint.pth')
        torch.save(checkpoint,path)

    
    def load_checkpoint(self,filepath):
        if os.path.isfile(filepath):
            checkpoint = torch.load(filepath)
            print("Loaded {} model trained with batch_size = {}, seen {} epochs and {} mini batches".
                format(self.runs_dir,checkpoint['batch_size'],checkpoint['epoch'],checkpoint['batch'])) 
            return checkpoint
        else:
            return None
        
            
    def load_model(self,inference_mode=True):
        filepath = join(self.runs_dir,'checkpoint.pth')
        checkpoint = self.load_checkpoint(filepath)
        
        model = self.model
        model.load_state_dict(checkpoint['state_dict'])
        
        if inference_mode:
            for parameter in model.parameters():
                parameter.requires_grad = False
            model.eval()
        return model

